File: CPP/vao.cpp
#include "vao.h"
#include "renderer.h"

// =================================================================================================
// "Premium version of" OpenGL vertex array objects. CVAO instances offer methods to convert python
// lists into the corresponding lists of OpenGL items (vertices, normals, texture coordinates, etc)
// The current implementation requires a fixed order of array buffer creation to comply with the 
// corresponding layout positions in the shaders implemented here.
// Currently offers shaders for cubemap and regular (2D) texturing.
// Implements loading of varying textures, so an application item derived from or using a CVAO instance
// (e.g. an ico sphere) can be reused by different other application items that require different 
// texturing. This implementation basically allows for reusing one single ico sphere instance whereever
// a sphere is needed.
// Supports indexed and non indexed vertex buffer objects.
//
// // Due to the current shader implementation (fixed position layout), buffers need to be passed in a
// fixed sequence: vertices, colors, ...
// TODO: Expand shader for all kinds of inputs (texture coordinates, normals)
// See also https://qastack.com.de/programming/8704801/glvertexattribpointer-clarification

void CVAO::Init (GLuint shape, CTexture* texture, CVector color) {
    m_shape = shape;
    glGenVertexArrays (1, &m_handle);
    m_texture = texture;
    m_color = color;
    m_minBrightness = 0.0f;
}


void CVAO::Destroy(void) {
    for (auto [i, vbo] : m_dataBuffers)
        vbo.Destroy();
    if (m_handle > 0) {
        Disable();
        glDeleteVertexArrays(1, &m_handle);
        m_handle = 0;
    }
}


CVAO& CVAO::Copy (CVAO const& other) {
    m_dataBuffers = other.m_dataBuffers;
    m_indexBuffer = other.m_indexBuffer;
    m_handle = other.m_handle;
    m_shape = other.m_shape;
    m_texture = other.m_texture;
    m_color = other.m_color;
    m_minBrightness = other.m_minBrightness;
    return *this;
}


// add a vertex or index data buffer
void CVAO::AddBuffer(const char* type, void * data, size_t dataSize, size_t componentType, size_t componentCount) {
    if (!strcmp(type, "Index"))
        AddIndexBuffer(data, dataSize, componentType);
    else
        AddVertexBuffer(type, data, dataSize, componentType, componentCount);
}


void CVAO::AddVertexBuffer(const char* type, void * data, size_t dataSize, size_t componentType, size_t componentCount) {
    CVBO* buffer = m_dataBuffers.Add (-1);
    buffer->Create(type, GL_ARRAY_BUFFER, int (m_dataBuffers.Length() - 1), data, dataSize, componentType, componentCount);
}


void CVAO::AddIndexBuffer(void * data, size_t dataSize, size_t componentType) {
    m_indexBuffer.Create("Index", GL_ELEMENT_ARRAY_BUFFER, -1, data, dataSize, componentType);
}


void CVAO::Render(bool useShader) {
    int shaderId = shaderHandler->SelectShader(useShader, EnableTexture());
    shaderHandler->Shader (shaderId).SetUniformVector("fillColor", m_color);
    if (shaderId == 0) {
        shaderHandler->Shader (shaderId).SetUniformFloat ("minBrightness", m_minBrightness);
    }
    Enable();
    if (!m_indexBuffer.m_data)
        glDrawArrays(m_shape, 0, m_dataBuffers[0].m_itemCount); // draw non indexed arrays
    else
        glDrawElements(m_shape, m_indexBuffer.m_itemCount, m_indexBuffer.m_componentType, nullptr); // draw using an index buffer
    Disable();
    if (shaderId > -1)
        glUseProgram(0);
    DisableTexture();
}

// =================================================================================================

File: Python/vao.py
import globals

from OpenGL import *
from OpenGL.GL import *
from shaders import *
from vbo import *
from vector import *
from texcoord import *

# =================================================================================================
# "Premium version of" OpenGL vertex array objects. CVAO instances offer methods to convert python
# lists into the corresponding lists of OpenGL items (vertices, normals, texture coordinates, etc)
# The current implementation requires a fixed order of array buffer creation to comply with the 
# corresponding layout positions in the shaders implemented here.
# Currently offers shaders for cubemap and regular (2D) texturing.
# Implements loading of varying textures, so an application item derived from or using a CVAO instance
# (e.g. an ico sphere) can be reused by different other application items that require different 
# texturing. This implementation basically allows for reusing one single ico sphere instance whereever
# a sphere is needed.
# Supports indexed and non indexed vertex buffer objects.
#
# # Due to the current shader implementation (fixed position layout), buffers need to be passed in a
# fixed sequence: vertices, colors, ...
# See also https://qastack.com.de/programming/8704801/glvertexattribpointer-clarification

class CVAO:
    def __init__ (self, shape, texture = None, color = CVector (1,1,1)):
        self.dataBuffers = []
        self.indexBuffer = None
        self.handle = glGenVertexArrays (1)
        self.shape = shape
        self.texture = texture
        self.color = color
        self.minBrightness = 0


    def Destroy (self):
        for vbo in self.dataBuffers:
            vbo.Destroy ()
        if (self.handle > 0):
            self.Disable ()
            glDeleteVertexArrays (1, [self.handle])
            self.handle = 0


    def Enable (self):
        glBindVertexArray (self.handle)


    def Disable (self):
        glBindVertexArray (0)


    # add a vertex or index data buffer
    def AddBuffer (self, type, data, dataSize, componentType, componentCount = 0):
        if (type == "Index"):
            self.AddIndexBuffer (data, dataSize, componentType)
        else:
            self.AddVertexBuffer (type, data, dataSize, componentType, componentCount)


    def AddVertexBuffer (self, type, data, dataSize, componentType, componentCount):
        buffer = CVBO (type, GL_ARRAY_BUFFER)
        buffer.Create (len (self.dataBuffers), data, dataSize, componentType, componentCount)
        self.dataBuffers.append (buffer)


    def AddIndexBuffer (self, data, dataSize, componentType):
        self.indexBuffer = CVBO ("Index", GL_ELEMENT_ARRAY_BUFFER)
        self.indexBuffer.Create (-1, data, dataSize, componentType)


    def SetTexture (self, texture):
        self.texture = texture


    def SetColor (self, color):
        self.color = color


    def SetMinBrightness (self, minBrightness):
        self.minBrightness = minBrightness


    def ResetMinBrightness (self):
        self.minBrightness = 0


    def EnableTexture (self):
        if (self.texture is None):
            return None
        self.texture.Enable ()
        return self.texture


    def DisableTexture (self):
        if (self.texture is not None):
            self.texture.Disable ()


    def Render (self, useShader = True):
        shaderId = globals.shaderHandler.SelectShader (useShader, self.EnableTexture ())
        globals.shaderHandler.shaders [shaderId].SetUniformVector ("fillColor", self.color)
        if (shaderId == 0):
            globals.shaderHandler.shaders [shaderId].SetUniformFloat ("minBrightness", self.minBrightness)
        self.Enable ()
        # glColor3f (self.color.x, self.color.y, self.color.z)
        if (self.indexBuffer is None):
            glDrawArrays (self.shape, 0, self.dataBuffers [0].itemCount) # draw non indexed arrays
        else:
            glDrawElements (self.shape, self.indexBuffer.size, self.indexBuffer.componentType, None) # draw using an index buffer
            self.Disable ()
        if (shaderId > -1):
            globals.shaderHandler.shaders [shaderId].Disable ()
        glUseProgram (0)
        self.DisableTexture ()


# =================================================================================================
